rumqttd/src/router/alertlog.rs

```diff
@@ -0,0 +1,244 @@
+use slab::Slab;
+
+use crate::protocol::{matches, Subscribe, Unsubscribe};
+use crate::router::FilterIdx;
+use crate::{ConnectionId, Filter, Offset, RouterConfig, Topic};
+
+use crate::segments::{CommitLog, Position};
+use crate::Storage;
+use std::collections::HashMap;
+use std::io;
+
+pub enum AlertTopic {
+    Connect,
+    Disconnect,
+    Subscribe,
+    Unsubscribe,
+}
+
+impl ToString for AlertTopic {
+    fn to_string(&self) -> String {
+        match self {
+            Self::Connect => String::from("/alerts/connect"),
+            Self::Disconnect => String::from("/alerts/disconnect"),
+            Self::Subscribe => String::from("/alerts/subscribe"),
+            Self::Unsubscribe => String::from("/alerts/unsubscribe"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Alert {
+    Connect(String),
+    Disconnect(String),
+    Subscribe(String, Subscribe),
+    Unsubscribe(String, Unsubscribe),
+}
+
+impl Alert {
+    // Topic prefix to which alerts are published to
+    pub fn topic_prefix(&self) -> String {
+        match self {
+            Self::Connect(_) => AlertTopic::Connect.to_string(),
+            Self::Disconnect(_) => AlertTopic::Disconnect.to_string(),
+            Self::Subscribe(_, _) => AlertTopic::Subscribe.to_string(),
+            Self::Unsubscribe(_, _) => AlertTopic::Unsubscribe.to_string(),
+        }
+    }
+
+    pub fn topic(&self) -> String {
+        match self {
+            Self::Connect(client_id) => {
+                format!("{}/{}", AlertTopic::Connect.to_string(), client_id)
```

Comment:
`Alert::topic()` may produce a filter of unexpected behavior for an unusual value of `client_id` instead of a topic

example: for `client_id = "+"`, `Alert::Connect(client_id).topic()` will be `alerts/connect/+`